src/inputs/TreeSelectField/TreeSelectField.tsx

```diff
@@ -0,0 +1,540 @@
+        ),
+      }));
+    }
+    reactToCollapse.current = true;
+  }, [collapsedKeys]);
+
+  // Update the filtered options when the input value changes
+  const onInputChange = useCallback(
+    (inputValue: string) => {
+      setFieldState((prevState) => {
+        return {
+          ...prevState,
+          inputValue,
+          allowCollapsing: inputValue.length === 0,
+          filteredOptions: prevState.allOptions.flatMap((o) =>
+            levelOptions(o, 0, inputValue.length > 0).filter(([option]) =>
+              contains(getOptionLabel(option), inputValue),
+            ),
+          ),
+        };
+      });
+    },
+    [setFieldState],
+  );
+
+  // Handle loading of the options in the case that they are loaded via a Promise.
+  const maybeInitLoad = useCallback(
+    async (
+      options: NestedOptionsOrLoad<O>,
+      fieldState: TreeFieldState<O>,
+      setFieldState: React.Dispatch<React.SetStateAction<TreeFieldState<O>>>,
+    ) => {
+      if (!Array.isArray(options)) {
+        setFieldState((prevState) => ({ ...prevState, optionsLoading: true }));
+        const loadedOptions = (await options.load()).options;
+        const filteredOptions = loadedOptions.flatMap((o) =>
+          levelOptions(o, 0, fieldState.inputValue.length > 0).filter(([option]) =>
+            contains(getOptionLabel(option), fieldState.inputValue),
+          ),
+        );
+
+        // Ensure the `unset` option is prepended to the top of the list if `unsetLabel` was provided
+        setFieldState((prevState) => ({
+          ...prevState,
+          filteredOptions,
+          allOptions: loadedOptions,
+          optionsLoading: false,
+        }));
+      }
+    },
+    [],
+  );
+
+  // Only on the first open of the listbox, we want to load the options if they haven't been loaded yet.
+  const firstOpen = useRef(true);
+  function onOpenChange(isOpen: boolean) {
+    if (!isOpen) debugger;
+    if (firstOpen.current && isOpen) {
+      maybeInitLoad(options, fieldState, setFieldState);
+      firstOpen.current = false;
+    }
+  }
+
+  // This is _always_ going to appear new. Maybe `useMemo`?
+  const comboBoxProps = {
+    ...otherProps,
+    placeholder: !values || values.length === 0 ? placeholder : "",
+    label: props.label,
+    inputValue: fieldState.inputValue,
+    // where we might want to do flatmap and return diff kind of array (children ? add level prop) inside children callback - can put markup wrapper div adds padding
+    // so we're not doing it multiple places
+    items: fieldState.filteredOptions,
+    isDisabled,
+    isReadOnly,
+    onInputChange,
+    onOpenChange,
+    children: ([item]: LeveledOption<O>) => (
+      // what we're telling it to render. look at padding here - dont have to pass down to tree option - filtered options is where we're flat mapping
+      <Item key={valueToKey(getOptionValue(item))} textValue={getOptionLabel(item)}>
```

Comment:
I don't remember where this landed, are we just trusting for now that values are tagged IDs and will always be unique?

Reply:
For now. I was going to have it be a follow up.